File: ios/Classes/PayByWalletPlugin.swift
import Flutter
import TerraPayWalletSDK
import UIKit

/// iOS half of the PayByWallet bridge.
///
/// Dart calls `launch` / `processPayment` down; the SDK's completion handler
/// callbacks are forwarded back up the same channel on the main thread.
public class PayByWalletPlugin: NSObject, FlutterPlugin {

  private static let channelName = "com.terrapay.paybywallet/sdk"

  private let channel: FlutterMethodChannel

  private init(channel: FlutterMethodChannel) {
    self.channel = channel
    super.init()
  }

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(
      name: channelName,
      binaryMessenger: registrar.messenger()
    )
    let instance = PayByWalletPlugin(channel: channel)
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  // MARK: - Host view controller
  //
  // Resolved lazily rather than captured at registration: under
  // UIApplicationSceneManifest (mandatory on iOS 26+ SDKs) the window and its
  // root FlutterViewController do not exist yet when plugins are registered.

  private var hostController: UIViewController? {
    let scenes = UIApplication.shared.connectedScenes.compactMap { $0 as? UIWindowScene }
    let window =
      scenes.first(where: { $0.activationState == .foregroundActive })?.keyWindow
      ?? scenes.first?.keyWindow
    var top = window?.rootViewController
    while let presented = top?.presentedViewController {
      top = presented
    }
    return top
  }

  // MARK: - Dart -> native

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "launch":
      launch(call, result: result)
    case "processPayment":
      processPayment(call, result: result)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func launch(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let args = call.arguments as? [String: Any] else {
      result(Self.error("INVALID_ARGUMENTS", "Expected a configuration map."))
      return
    }
    guard let controller = hostController else {
      result(Self.error("INVALID_CONTEXT", "No visible view controller to present from."))
      return
    }

    // The iOS SDK takes the balance as a string; Dart sends a number.
    let walletBalance: String?
    switch args["walletBalance"] {
    case let value as NSNumber: walletBalance = value.stringValue
    case let value as String: walletBalance = value
    default: walletBalance = nil
    }

    let environment: TPEnvironment =
      (args["environment"] as? String) == "production" ? .production : .sandbox

    let config = TerraPayWalletSDKConfig(
      controller: controller,
      accessToken: args["accessToken"] as? String ?? "",
      refreshToken: args["refreshToken"] as? String ?? "",
      subscriberDialCode: args["subscriberDialCode"] as? String ?? "",
      subscriberCountry: args["subscriberCountry"] as? String ?? "",
      subscriberCountryName: args["subscriberCountryName"] as? String ?? "",
      subscriberName: args["subscriberName"] as? String ?? "",
      subscriberMSISDN: args["subscriberMSISDN"] as? String ?? "",
      subscriberCurrency: args["subscriberCurrency"] as? String ?? "",
      walletBalance: walletBalance,
      primaryColor: args["primaryColor"] as? String ?? "",
      secondaryColor: args["secondaryColor"] as? String ?? "",
      environment: environment
    )

    TerraPayWalletClient.shared.launch(with: config) { [weak self] type, error, merchant, status in
      self?.forward(type: type, error: error, merchant: merchant, status: status)
    }
    result(nil)
  }

  private func processPayment(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard
      let args = call.arguments as? [String: Any],
      let transactionId = args["transactionId"] as? String,
      !transactionId.isEmpty
    else {
      result(Self.error("INVALID_TRANSACTION_ID", "transactionId is required."))
      return
    }
    TerraPayWalletClient.shared.processPayment(
      controller: hostController,
      transactionId: transactionId
    )
    result(nil)
  }

  // MARK: - Native -> Dart

  private func forward(
    type: TPLaunchType,
    error: TPErrorInfo?,
    merchant: TPMerchant?,
    status: TPPaymentStatus?
  ) {
    switch type {
    case .onPINAuthenticate:
      send("onPinAuthenticate", [
        "merchantName": merchant?.merchantName as Any,
        "subscriberAmount": merchant?.subscriberAmount as Any,
        "subscriberCurrency": merchant?.subscriberCurrency as Any,
      ])
    case .onPaymentSuccess:
      send("onPaymentSuccess", ["raw": status?.description as Any])
    case .onPaymentFailure:
      send("onPaymentFailure", [
        "raw": status?.description as Any,
        "responseCode": error?.code as Any,
        "responseMessage": error?.message as Any,
      ])
    case .onError:
      send("onError", [
        "code": error?.code ?? "",
        "message": error?.message ?? "Something went wrong.",
      ])
    case .cancelled:
      send("onCancelled", [
        "code": error?.code ?? "",
        "message": error?.message ?? "User cancelled.",
      ])
    case .closed:
      send("onClosed", [:])
    @unknown default:
      send("onError", ["code": "", "message": "Unrecognised SDK result."])
    }
  }

  private func send(_ method: String, _ arguments: [String: Any]) {
    // Strip NSNull so optional fields arrive as Dart nulls.
    let cleaned = arguments.filter { !($0.value is NSNull) }
    DispatchQueue.main.async { [weak self] in
      self?.channel.invokeMethod(method, arguments: cleaned)
    }
  }

  private static func error(_ code: String, _ message: String) -> FlutterError {
    FlutterError(code: code, message: message, details: nil)
  }
}

File: android/src/main/kotlin/com/terrapay/paybywallet_flutter/PayByWalletPlugin.kt
package com.terrapay.paybywallet_flutter

import android.app.Activity
import android.os.Handler
import android.os.Looper
import androidx.activity.ComponentActivity
import com.terrapay.payByWallet.network.MerchantDetailsModel
import com.terrapay.payByWallet.network.PaymentResponse
import com.terrapay.payByWallet.network.TerraPayClient
import com.terrapay.payByWallet.network.TerraPayConfig
import com.terrapay.payByWallet.network.TerraPayResult
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/**
 * Android half of the PayByWallet bridge.
 *
 * Dart calls `launch` / `processPayment` down; the SDK's [TerraPayResult]
 * callbacks are forwarded back up the same channel on the main thread.
 */
class PayByWalletPlugin :
    FlutterPlugin, ActivityAware, MethodChannel.MethodCallHandler {

    private companion object {
        const val CHANNEL = "com.terrapay.paybywallet/sdk"
    }

    private lateinit var channel: MethodChannel
    private var activity: Activity? = null
    private val main = Handler(Looper.getMainLooper())

    // ---- Lifecycle ----------------------------------------------------------

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, CHANNEL)
        channel.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onDetachedFromActivity() {
        activity = null
    }

    // ---- Dart -> native -----------------------------------------------------

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "launch" -> launch(call, result)
            "processPayment" -> processPayment(call, result)
            else -> result.notImplemented()
        }
    }

    /**
     * The SDK's UI is Compose-based and it rejects any context that is not a
     * [ComponentActivity]. Flutter's default `FlutterActivity` extends plain
     * `android.app.Activity`, so the host app must use `FlutterFragmentActivity`.
     * Checking here turns an opaque SDK error into an actionable one.
     */
    private fun requireComponentActivity(result: MethodChannel.Result): ComponentActivity? {
        val current = activity
        if (current is ComponentActivity) return current
        result.error(
            "INVALID_CONTEXT",
            "PayByWallet needs a ComponentActivity. Make your MainActivity extend " +
                "io.flutter.embedding.android.FlutterFragmentActivity instead of FlutterActivity.",
            null,
        )
        return null
    }

    private fun launch(call: MethodCall, result: MethodChannel.Result) {
        val activity = requireComponentActivity(result) ?: return

        val balance = when (val raw = call.argument<Any>("walletBalance")) {
            is Number -> raw.toDouble()
            is String -> raw.toDoubleOrNull()
            else -> null
        }
        if (balance == null) {
            result.error("INVALID_WALLET_BALANCE", "walletBalance must be a number.", null)
            return
        }

        val config = TerraPayConfig(
            primaryColor = call.argument<String>("primaryColor").orEmpty(),
            secondaryColor = call.argument<String>("secondaryColor").orEmpty(),
            subscriberDialCode = call.argument<String>("subscriberDialCode").orEmpty(),
            subscriberMSISDN = call.argument<String>("subscriberMSISDN").orEmpty(),
            walletBalance = balance,
            subscriberCountryName = call.argument<String>("subscriberCountryName").orEmpty(),
            subscriberCountry = call.argument<String>("subscriberCountry").orEmpty(),
            subscriberCurrency = call.argument<String>("subscriberCurrency").orEmpty(),
            subscriberName = call.argument<String>("subscriberName").orEmpty(),
            accessToken = call.argument<String>("accessToken").orEmpty(),
            refreshToken = call.argument<String>("refreshToken").orEmpty(),
        )

        // Surface configuration problems as a failed call rather than as an
        // asynchronous onError, so Dart can await the launch.
        TerraPayClient.validateInputFields(activity, config)?.let { error ->
            result.error(error.code, error.message, null)
            return
        }

        TerraPayClient.init(
            context = activity,
            config = config,
            referenceNumber = call.argument<String>("referenceNumber"),
            terraPayResult = callbacks,
        )
        result.success(null)
    }

    private fun processPayment(call: MethodCall, result: MethodChannel.Result) {
        val activity = requireComponentActivity(result) ?: return
        val transactionId = call.argument<String>("transactionId")
        if (transactionId.isNullOrBlank()) {
            result.error("INVALID_TRANSACTION_ID", "transactionId is required.", null)
            return
        }
        TerraPayClient.processPayment(context = activity, transactionId = transactionId)
        result.success(null)
    }

    // ---- Native -> Dart -----------------------------------------------------

    private fun send(method: String, args: Map<String, Any?>) {
        main.post { channel.invokeMethod(method, args) }
    }

    private val callbacks = object : TerraPayResult {

        override fun onPinAuthenticate(merchantDetails: MerchantDetailsModel) {
            send(
                "onPinAuthenticate",
                mapOf(
                    "merchantName" to merchantDetails.merchantName,
                    "subscriberAmount" to merchantDetails.subscriberAmount,
                    "subscriberCurrency" to merchantDetails.subscriberCurrency,
                ),
            )
        }

        override fun onPaymentSuccess(paymentResponse: PaymentResponse) =
            send("onPaymentSuccess", paymentResponse.toMap())

        override fun onPaymentFailure(paymentResponse: PaymentResponse) =
            send("onPaymentFailure", paymentResponse.toMap())

        override fun onError(errorCode: String, message: String) =
            send("onError", mapOf("code" to errorCode, "message" to message))

        override fun onCancelled(errorCode: String, message: String) =
            send("onCancelled", mapOf("code" to errorCode, "message" to message))
    }

    private fun PaymentResponse.toMap(): Map<String, Any?> = mapOf(
        "responseCode" to responseCode,
        "responseMessage" to responseMessage,
        "gatewayReferenceId" to gatewayReferenceId,
        "orderId" to orderId,
    )
}
